driver/src/contracts/stablex_contract.rs

```diff
@@ -59,12 +59,84 @@ impl StableXContract for StableXContractImpl {
 
     fn submit_solution(
         &self,
-        _batch_index: U256,
-        _orders: Vec<Order>,
-        _solution: Solution,
+        batch_index: U256,
+        orders: Vec<Order>,
+        solution: Solution,
     ) -> Result<()> {
-        unimplemented!();
+        let account = self
+            .base
+            .account_with_sufficient_balance()
+            .ok_or("Not enough balance to send Txs")?;
+        let (prices, token_ids_for_price) = encode_prices_for_contract(solution.prices);
+        let (owners, order_ids, volumes) =
+            encode_execution_for_contract(orders, solution.executed_sell_amounts);
+
+        self.base
+            .contract
+            .call(
+                "submitSolution",
+                (
+                    batch_index,
+                    owners,
+                    order_ids,
+                    volumes,
+                    prices,
+                    token_ids_for_price,
+                ),
+                account,
+                Options::default(),
+            )
+            .wait()
+            .map_err(DriverError::from)
+            .map(|_| ())
+    }
+}
+
+fn encode_prices_for_contract(price_vector: Vec<u128>) -> (Vec<U128>, Vec<U128>) {
+    // Representing the solution's price vector more compactly as:
+    // sorted_touched_token_ids, non_zero_prices which are logically bound by index.
+    // Example solution.prices = [3, 0, 1] will be transformed into [0, 2], [3, 1]
+    let mut ordered_token_ids: Vec<U128> = vec![];
+    let mut prices: Vec<U128> = vec![];
+    for (token_id, price) in price_vector.into_iter().enumerate() {
+        if price > 0 {
+            ordered_token_ids.push(U128::from(token_id));
+            prices.push(U128::from(price.to_be_bytes()));
+        }
+    }
+    (prices, ordered_token_ids)
+}
+
+fn encode_execution_for_contract(
+    orders: Vec<Order>,
+    executed_sell_amounts: Vec<u128>,
+) -> (Vec<H160>, Vec<U128>, Vec<U128>) {
+    assert_eq!(
+        orders.len(),
+        executed_sell_amounts.len(),
+        "Received inconsistent auction result data."
+    );
+    // Note that buy_amounts do not play a role here since all orders are sell orders.
+    let mut owners: Vec<H160> = vec![];
+    let mut order_ids: Vec<U128> = vec![];
+    let mut volumes: Vec<U128> = vec![];
+    for (order_index, sell_amount) in executed_sell_amounts.into_iter().enumerate() {
+        if sell_amount > 0 {
```

Comment:
If you apply a filter function before iterating, you reduce the cyclomatic complexity